### PR TITLE
[QAOA] Adds .map_wires Method to qml.Hamiltonian Class

### PR DESCRIPTION
**Pull request by Lucaman99**
**Note:** The purpose of this PR has changed, please read new description and ignore changes in old commits.

**Context:**

**Description of the Change:**

Adds a new method to the `qml.Hamiltonian` class that maps a Hamiltonian to a new set of wires. For instance, if we define the following Hamiltonian:

``` python

import pennylane as qml

coeffs = [1, 1]
ops = [qml.PauliZ(0), qml.PauliX(0) @ qml.PauliX(1)]

hamiltonian = qml.Hamiltonian(coeffs, ops)

```

we will then have:

```
new_hamiltonian = hamiltonian.map_wires({0:3, 1:7})
```

```
>>> print(hamiltonian)
(1.0) [Z3] + (1.0) [X3 X7]
```

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/vqe/vqe.py [hunk 1]
```diff
@@ -120,6 +120,32 @@ def __str__(self):
...
+            bool: ``True`` if the Hamiltonian is diagonal in the computational basis, ``False`` otherwise
+        """
+
+        diagonals = ["PauliZ", "Identity"]
+
+        non_diagonal_coeffs = []
+        non_diagonal_gates = []
+
+        for i, term in enumerate(self._ops):
+            gates = [term.name] if isinstance(term.name, str) else term.name
```
Comment (ixfoduap) [inline pennylane/vqe/vqe.py:137]: Not really gates, right? Maybe observables or operators is more clear?

File: pennylane/vqe/vqe.py [hunk 2]
```diff
@@ -120,6 +120,32 @@ def __str__(self):
...
+            for j in gates:
```
Comment (ixfoduap) [inline pennylane/vqe/vqe.py:138]: Maybe `for g in gates`? 

It's important to make code easily-understandable, in case we need to revisit it in the future

File: pennylane/vqe/vqe.py [hunk 3]
```diff
@@ -120,6 +120,32 @@ def __str__(self):
...
+                if j not in diagonals:
+                    if gates not in non_diagonal_gates:
```
Comment (ixfoduap) [inline pennylane/vqe/vqe.py:140]: This double negation is a bit awkward. Can we reformulate? Something like `if gates in diagonal_gates`?

File: pennylane/vqe/vqe.py [hunk 4]
```diff
@@ -120,6 +120,32 @@ def __str__(self):
...
+                        non_diagonal_gates.append(gates)
+                        non_diagonal_coeffs.append(self._coeffs[i])
+                    else:
+                        non_diagonal_coeffs[non_diagonal_gates.index(gates)] += self._coeffs[i]
+                    break
+
+        return all([not bool(i) for i in non_diagonal_coeffs]) or len(non_diagonal_coeffs) == 0
```
Comment (ixfoduap) [inline pennylane/vqe/vqe.py:147]: So what happens here if the non-diagonal coefficients are very small but non-zero?

Comment (Lucaman99) [reply pennylane/vqe/vqe.py:147]: The function would return `False`. I'm assuming you're concerned about floating-point errors? If so, we can set some kind of tolerance 🙂

File: pennylane/vqe/vqe.py [hunk 5]
```diff
@@ -120,6 +120,32 @@ def __str__(self):
 
         return "\n+ ".join(terms)
 
+    @property
```
[the remark this answers is not in the document]
Comment (Lucaman99) [reply pennylane/vqe/vqe.py:123]: If you are referring to the `is_diagonal` method as a whole, it is an attribute of the pennylane.Hamiltonian object, so this is really the only place for it

File: pennylane/vqe/vqe.py (hunk 4 above)
Comment (Lucaman99) [reply pennylane/vqe/vqe.py:147]: Done!

File: pennylane/vqe/vqe.py [hunk 6]
```diff
@@ -120,6 +120,35 @@ def __str__(self):
...
+    def is_diagonal(self):
+        r""" Checks if a Hamiltonian is diagonal in the computational basis.
+
+        Returns:
+            bool: ``True`` if the Hamiltonian is diagonal in the computational basis, ``False`` otherwise
```
Comment (josh146) [inline pennylane/vqe/vqe.py:128]:
```suggestion
            bool: ``True`` if the Hamiltonian is diagonal in the computational basis, ``False`` otherwise.
```

File: pennylane/vqe/vqe.py [hunk 7]
```diff
@@ -120,6 +120,35 @@ def __str__(self):
 
         return "\n+ ".join(terms)
 
+    @property
```
Comment (josh146) [inline pennylane/vqe/vqe.py:123]: @Lucaman99 I _might_ suggest making this a method rather than a property. I usually expect properties not to perform (expensive) logic with every call; a method better indicates to the user that logic could be occurring with every call.

If a property is expensive to calculate, or has side effects, then it likely would be much better as a normal method. Making it look like an attribute access means that developers will likely think of it as an attribute access, which is cheap and non-destructive; this would be a rich source of bugs or performance problems.

File: pennylane/vqe/vqe.py [hunk 8]
```diff
@@ -120,6 +120,35 @@ def __str__(self):
...
+        for i, term in enumerate(self._ops):
+            gates = [term.name] if isinstance(term.name, str) else term.name
+            for g in gates:
+                if g not in diagonals:
+                    if gates in non_diagonal_obs:
+                        non_diagonal_coeffs[non_diagonal_obs.index(gates)] += self._coeffs[i]
+                    else:
+                        non_diagonal_obs.append(gates)
+                        non_diagonal_coeffs.append(self._coeffs[i])
+                    break
```
Comment (josh146) [inline pennylane/vqe/vqe.py:145]: @Lucaman99 it might be worth including line comments here explaining the logic, I find it a bit hard to follow at first read.

**Comment by Lucaman99** [comment]
@josh146 @ixfoduap I just realized that there are massive issues with the current implementation of the ``is_diagonal`` function, as I wasn't taking into account terms containing custom ``Hermitian`` observables. I'm in the process of fixing it now!

**Comment by Lucaman99** [comment]
Ok, it's all fixed now! 😄

File: pennylane/vqe/vqe.py [hunk 9]
```diff
@@ -120,6 +120,39 @@ def __str__(self):
...
+        non_diagonal_obs = []
+
+        # Loops through each term of the Hamiltonian
+        for i, term in enumerate(self._ops):
+
+            # Prunes identity from all tensor products
+            term = term.prune() if isinstance(term, qml.operation.Tensor) else term
+
+            # Combines like terms (provided they are non-diagonal)
+            if bool(np.count_nonzero(term.matrix - np.diag(np.diagonal(term.matrix)))):
```
Comment (josh146) [inline pennylane/vqe/vqe.py:140]: @Lucaman99 nice! And it will generalize to numeric observables as well.

However, a thought crosses my mind. This approach will work for `qml.Hermitian`, but calling `term.matrix` will be _very_ expensive for Hamiltonians on a large number of wires. Your other approach, where terms were simply matched against `qml.PauliZ` and `qml.Identity` scaled significantly better.

Would it be possible to only call `.matrix` on the `qml.Hermitian` terms directly, and continue to pattern match against `{qml.PauliZ, qml.Identity}` for the remaining observables?

Another approach would be to branch between this method and the old method if `qml.Hermitian` is in the term or not.

File: pennylane/vqe/vqe.py [hunk 10]
```diff
@@ -120,6 +120,39 @@ def __str__(self):
...
+
+                t = [term.matrix.tolist(), term.wires]
+
+                if t in non_diagonal_obs:
+                    non_diagonal_coeffs[non_diagonal_obs.index(t)] += self._coeffs[i]
+                else:
+                    non_diagonal_obs.append(t)
+                    non_diagonal_coeffs.append(self._coeffs[i])
```
Comment (josh146) [inline pennylane/vqe/vqe.py:148]: This sounds like it could be simplified by using a dictionary for `non_diagonal_obs`/`non_diagonal_coeffs`, since the order doesn't matter :)

File: tests/test_vqe.py [hunk 11]
```diff
@@ -184,6 +184,23 @@ def test_hamiltonian_valid_init(self, coeffs, ops):
...
 
+    @pytest.mark.parametrize(
+        ("coeffs", "ops", "is_diagonal"),
+        [
+            ([1, 1, 1], [qml.PauliZ(0), qml.PauliZ(1), qml.PauliZ(2)], True),
+            ([1, 1, 1], [qml.PauliZ(0) @ qml.PauliZ(1), qml.PauliZ(1), qml.PauliZ(2)], True),
+            ([1, 1, 1], [qml.PauliZ(0), qml.PauliZ(1), qml.PauliX(2)], False),
+            ([1, -1, 1], [qml.PauliX(0) @ qml.PauliX(1), qml.PauliX(0) @ qml.PauliX(1), qml.Identity(2)], True),
+            ([1, -1, 2], [qml.PauliX(0) @ qml.PauliZ(1), qml.PauliX(0) @ qml.PauliZ(1) @ qml.Identity(2), qml.Identity(2)], True),
+            ([1, -1, 2], [qml.Hermitian(np.array([[0, 1], [1, 0]]), 0) @ qml.Identity(1), qml.PauliX(0), qml.PauliZ(1)], True)
```
Comment (josh146) [inline tests/test_vqe.py:195]: Would also be good to include a test case with `qml.Hermitian` that is diagonal!

File: pennylane/vqe/vqe.py (hunk 9 above)
Comment (Lucaman99) [reply pennylane/vqe/vqe.py:140]: Yes, I think this should be possible! One thing I was concerned about was the edge case in which the user defines a Hermitian that is equivalent to a Pauli gate. For instance:

``` python
H = qml.Hamiltonian([1, 1, -1], [qml.PauliZ(0), qml.Hermitian(np.array([[0, 1], [1, 0]), 1), qml.PauliX(1)])
print(H.is_diagonal())
```

should return `True`. However, I think I may have a solution! I believe there is a utility function that decomposes Hermitians into Pauli components, so we could use that, re-define the Hamiltonian, and then use the old logic to pattern match against diagonal gates. What do you think?

Comment (Lucaman99) [reply pennylane/vqe/vqe.py:140]: I know the decomposition would be very expensive, but hopefully, cases in which the user is defining Hamiltonians in terms of massive, custom Hermitians is rare 😆

**Comment by Lucaman99** [comment]
This is proving to be **highly** non-trivial 😆

File: pennylane/vqe/vqe.py (hunk 9 above)
Comment (josh146) [reply pennylane/vqe/vqe.py:140]: I think this is a good plan! The thing I was more worried about was if the user has a Hamiltonian of the form

```python
qml.PauliZ(0) @ qml.PauliZ(1)  @ ...  @ qml.PauliZ(100) @ qml.PauliZ(101) @ qml.PauliZ(102)
```
etc. This is trivial to determine if it's diagonal using the operator matching approach, but will be a _huge_ computational cost to instead compute the dense matrix!

**Comment by josh146** [comment]
> This is proving to be highly non-trivial 😆
@Lucaman99 how so?

**Comment by Lucaman99** [comment]
> > This is proving to be highly non-trivial 😆
> > @Lucaman99 how so?

@josh146 In order to account for all edge cases, where a user may supply a custom Hermitian term that cancels with an off-diagonal Pauli observable, we need two very complicated helper methods that decompose the Hermitian in terms of Paulis, re-constructs the new, decomposed Hamiltonian, and then performs the is_diagonal check.

@ixfoduap thinks that maybe we should stop working on this feature, and instead focus on implementing a more basic check when the QAOA cost layer is being built (to ensure the user-supplied cost Hamiltonian is written exclusively in terms if diagonal terms). I agree, for the most part, however, I'm pretty much done implementing the complicated functions, and I think they may be useful outside the context of the `is_diagonal` method anyways.

What do you think?

**Comment by co9olguy** [comment]
@Lucaman99 I'm curious if the title and description of this PR still match the intention. I got a bit confused when browsing through what the goal was

**Comment by Lucaman99** [comment]
@co9olguy Yes, the purpose of this PR pivoted earlier today, I'm making changes now!

**Comment by josh146** [comment]
@Lucaman99 in future, always better to close an existing PR, and create a new branch/PR if the scope changes significantly 🙂

**Comment by Lucaman99** [comment]
@josh146 You're right, I'll keep that in mind in the future 👍

File: pennylane/vqe/vqe.py [hunk 12]
```diff
@@ -120,6 +120,52 @@ def __str__(self):
...
+
+        for term in self.ops:
+
+            tensor_term = []
+            term = qml.operation.Tensor(term) if isinstance(term.name, str) else term
+
+            for i in term.obs:
+
+                name = i.name
+                new_wires = list(map(f, i.wires))
```
Comment (josh146) [inline pennylane/vqe/vqe.py:158]: Instead of using `list` and `map`, you could perhaps instead do
```suggestion
                new_wires = [wire_map.get(wire, wire) for wire in i.wires]
```

where `wire_map.get(key, default)` simply returns `default` if the key doesn't exist. This is slightly more 'pythonic' than the functional approach

File: pennylane/vqe/vqe.py [hunk 13]
```diff
@@ -120,6 +120,52 @@ def __str__(self):
...
+
+                if name == "Hermitian":
+                    tensor_term.append(obs[name](i.matrix, wires=new_wires))
+                else:
+                    tensor_term.append(obs[name](wires=new_wires))
```
Comment (josh146) [inline pennylane/vqe/vqe.py:163]: Here, you could instead do
```suggestion
                if name == "Hermitian":
                    tensor_term.append(qml.Hermitian(i.matrix, wires=new_wires))
                else:
                    tensor_term.append(i.__class__(wires=new_wires))
```

This avoids the need to create the `obs` dictionary, which might be hard to maintain if the observables change in other parts of the library.

Comment (josh146) [reply pennylane/vqe/vqe.py:163]: In fact, you can probably also drop the if statement:

```suggestion
                tensor_term.append(i.__class__(*i.data, wires=new_wires))
```

File: pennylane/vqe/vqe.py [hunk 14]
```diff
@@ -120,6 +120,52 @@ def __str__(self):
 
         return "\n+ ".join(terms)
 
+    def map_wires(self, wire_map):
+        r""" Maps the wires of a Hamiltonian to a new set of wires
```
Comment (josh146) [inline pennylane/vqe/vqe.py:124]:
```suggestion
        r""" Maps the wires of a Hamiltonian to a new set of wires.
```

File: pennylane/vqe/vqe.py [hunk 15]
```diff
@@ -120,6 +120,52 @@ def __str__(self):
...
+
+        Args:
+           wire_map (dict) : A dictionary that assigns new wires to a subset of the wires tracked in the Hamiltonian
+
+        Returns:
+            ~.Hamiltonian:
```
Comment (josh146) [inline pennylane/vqe/vqe.py:130]:
```suggestion
            .Hamiltonian:
```

File: pennylane/vqe/vqe.py [hunk 16]
```diff
@@ -120,6 +120,52 @@ def __str__(self):
...
+        """
```
Comment (josh146) [inline pennylane/vqe/vqe.py:131]: It would be good to include a small usage example here :)

File: tests/test_vqe.py [hunk 17]
```diff
@@ -184,14 +188,56 @@ def test_hamiltonian_valid_init(self, coeffs, ops):
...
+                    ],
+                ),
+                {0: 5, 1: 3},
+                [1, 2, 1],
+                [["PauliX", "PauliZ"], ["Hermitian", "PauliZ"], ["PauliX"]],
+                [[5, 3], [3, 5], [4]],
+            ),
+        ],
+    )
+    def test_hamiltonian_map_wires(self, hamiltonian, map, new_coeffs, new_ops, new_wires):
```
Comment (josh146) [inline tests/test_vqe.py:224]: What does this test check? Good practice to include a docstring explaining what it is testing/validating!

**Comment by co9olguy** [comment]
I'd really recommend here not implementing this feature. 
There are two reasons for this recommendation:
- It's not a make-or-break feature for users to be able to do this
- We want to support wire maps at the lowest level (i.e., the operators themselves). When that is implemented, it's a simple extension to support on objects that contain operators as attributes
